vllm/model_executor/layers/fused_moe/deepep_hybrid_prepare_finalize.py

```diff
@@ -0,0 +1,280 @@
+
+
+class DeepEPHybridPrepareAndFinalize(mk.FusedMoEPrepareAndFinalize):
+    """
+    Prepare/Finalize using DeepEP Hybrid kernels.
+    """
+
+    @staticmethod
+    def maybe_roundup_layer_hidden_size(hidden_size: int, dtype: torch.dtype) -> int:
+        # Round up hidden size so it is compatible with DeepEP Hybrid
+        # kernels.
+        alignment = 128
+        return round_up(hidden_size, alignment)
+
+    def __init__(
+        self,
+        buffer: "deep_ep.HybridEPBuffer",
+        num_dispatchers: int,
+        rank_expert_offset: int,
+        num_local_experts: int,
+    ):
+        super().__init__()
+        self.buffer = buffer
+        self.num_dispatchers_ = num_dispatchers
+        self.rank_expert_offset = rank_expert_offset
+        self.handle = None
+        self.expert_probs = None
+        self.num_local_experts = num_local_experts
+        # Set to True to try cudagraph support.
+        self.do_routing = False
+
+    def num_dispatchers(self) -> int:
+        return self.num_dispatchers_
+
+    def output_is_reduced(self) -> bool:
+        return True
+
+    @property
+    def activation_format(self) -> mk.FusedMoEActivationFormat:
+        return mk.FusedMoEActivationFormat.Standard
+
+    def max_num_tokens_per_rank(self) -> int | None:
+        return None
+
+    def topk_indices_dtype(self) -> torch.dtype | None:
+        return torch.int64
+
+    def supports_async(self) -> bool:
+        return False
+
+    # Copied from DeepEP hybrid_ep_buffer.py
+    def construct_handle(
+        self,
+        hidden_dim: int,
+        num_of_tokens: int,
+        num_of_experts: int,
+        routing_map: torch.Tensor,
+    ):
+        config = self.buffer.update_template_config(
+            hidden_dim=hidden_dim,
+            max_num_of_tokens_per_rank=num_of_tokens,
+        )
+
+        global_routing_map = get_dp_group().all_gather(routing_map, dim=0)
+
+        # Run the metadata preprocessing kernel.
+        (
+            sparse_to_dense_map,
+            rdma_to_attn_map,
+            attn_to_rdma_map,
+            num_dispatched_tokens_tensor,
+            local_expert_routing_map,
+        ) = self.buffer.runtime.metadata_preprocessing(
+            config=config,
+            routing_map=global_routing_map,
+            num_of_tokens_per_rank=num_of_tokens,
+        )
+
+        # Create the handle using the data generated by the preprocessing kernel.
+        return (
+            sparse_to_dense_map,
+            rdma_to_attn_map,
+            attn_to_rdma_map,
+            num_dispatched_tokens_tensor,
+            local_expert_routing_map,
+            num_of_tokens,
+            config,
+        )
+
+    def create_new_topk_data(
+        self,
+        topk_ids: torch.Tensor,
+        topk_weights: torch.Tensor,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        all_topk_ids, all_topk_weights = get_dp_group().all_gatherv(
+            [topk_ids, topk_weights], dim=0
+        )
```

Comment:
We're doing a dense `all_gatherv` after calling `dispatch`? This is not going to scale well. Could you explain why we need to do this?

Reply:
I've tried using expert_probs but I could not find a way to reconstruct the ordering so the results are gibberish.